hpy/devel/src/runtime/buildvalue.c

```diff
@@ -0,0 +1,145 @@
+ * format string.
+ *
+ * Building complex values with HPy_BuildValue is more convenient than the equivalent
+ * code that uses more granular APIs with proper error handling and cleanup. Moreover,
+ * HPy_BuildValue provides straightforward way to port existing code that uses
+ * Py_BuildValue.
+ *
+ * Supported Formatting Strings
+ * ----------------------------
+ *
+ * Numbers
+ * ~~~~~~~
+ *
+ * ``i (int) [int]``
+ *     Convert a plain C int to a Python integer object.
+ *
+ * ``f (float) [float]``
+ *     Convert a C float to a Python floating point number.
+ *
+ * Collections
+ * ~~~~~~~
+ *
+ * ``(items) (tuple) [matching-items]``
+ *     Convert a sequence of C values to a Python tuple with the same number of items.
+ *
+ * API
+ * ---
+ *
+ */
+
+#include "hpy.h"
+#include <stdarg.h>
+
+static HPy_ssize_t count_items(HPyContext *ctx, const char *fmt, char end);
+static HPy build_tuple(HPyContext *ctx, const char **fmt, va_list *values, HPy_ssize_t size, char expected_end);
+static HPy build_single(HPyContext *ctx, const char **fmt, va_list *values);
+
+HPyAPI_HELPER
+HPy HPy_BuildValue(HPyContext *ctx, const char *fmt, ...)
+{
+    va_list values;
+    va_start(values, fmt);
+    HPy_ssize_t size = count_items(ctx, fmt, '\0');
+    if (size < 0) {
+        return HPy_NULL;
+    } else if (size == 1) {
+        return build_single(ctx, &fmt, &values);
+    } else {
+        return build_tuple(ctx, &fmt, &values, size, '\0');
+    }
+}
+
+static HPy_ssize_t count_items(HPyContext *ctx, const char *fmt, char end)
+{
+    HPy_ssize_t level = 0, result = 0;
+    while (*fmt != end) {
+        switch (*fmt++) {
+            case '\0':
+                // Premature end
+                HPyErr_SetString(ctx, ctx->h_SystemError, "unmatched paren in format");
+                return -1;
+
+            case '[':
+            case '(':
+            case '{':
+                if (level == 0) {
+                    result++;
+                }
+                level++;
+                break;
+
+            case ']':
+            case ')':
+            case '}':
+                level--;
+                break;
+
+            case ' ':
+                break;
+
+            default:
+                if (level == 0) {
+                    result++;
+                }
+        }
+    }
+    return result;
+}
+
+static HPy build_single(HPyContext *ctx, const char **fmt, va_list *values)
+{
+    switch (*(*fmt)++) {
+        case '(': {
+            HPy_ssize_t size = count_items(ctx, *fmt, ')');
+            if (size < 0) {
+                return HPy_NULL;
+            }
+            return build_tuple(ctx, fmt, values, size, ')');
+        }
+
+        case 'i':
+            return HPyLong_FromLong(ctx, (long)va_arg(*values, int));
+
+        case 'f':
+            return HPyFloat_FromDouble(ctx, (double)va_arg(*values, double));
+
+        default: {
+            HPyErr_SetString(ctx, ctx->h_SystemError, "NULL object passed to Py_BuildValue");
```

Comment:
I think the error message should be `"bad format char passed to HPy_BuildValue"`